File: testers/pg/java/src/java/adventureworks/WithConnection.java
package adventureworks;

import dev.typr.foundations.Connection;
import dev.typr.foundations.ConnectionRead;
import dev.typr.foundations.SqlConsumer;
import dev.typr.foundations.SqlFunction;
import dev.typr.foundations.Transactor;
import dev.typr.foundations.connect.PgConfig;

public class WithConnection {
  private static final PgConfig CONFIG =
      PgConfig.builder("localhost", 6432, "Adventureworks", "postgres", "password").build();

  private static final Transactor TRANSACTOR = Transactor.create(CONFIG).rollbackOnly();

  public static <T> T apply(SqlFunction<Connection, T> f) {
    return TRANSACTOR.transact(f);
  }

  public static void run(SqlConsumer<Connection> f) {
    TRANSACTOR.transactVoid(f);
  }

  public static <T> T applyRead(SqlFunction<ConnectionRead, T> f) {
    return TRANSACTOR.transactRead(f);
  }
}

File: testers/pg/scala/javatypes/src/scala/adventureworks/WithConnection.scala
package adventureworks

import dev.typr.foundations.{Connection, SqlFunction, Transactor}
import dev.typr.foundations.connect.PgConfig

object WithConnection {
  private val config = PgConfig.builder("localhost", 6432, "Adventureworks", "postgres", "password").build()
  private val transactor = Transactor.create(config).rollbackOnly()

  def apply[T](f: Connection ?=> T): T = {
    val fn: SqlFunction[Connection, T] = c => f(using c)
    transactor.transact(fn)
  }
}
